### PR TITLE
Use newlib instead of libnustd.

### PR DESCRIPTION
**Pull request by GiacomoGarbin**
TODO: Remove libnustd from setups as well.

### CODE REVIEW

**Comment by mwpenny** [comment]
Nice. And thank you for updating the scripts + docs as well.

I'm going to do some testing to see how this affects performance - namely `sinf()` and `cosf()`.

**Comment by GiacomoGarbin** [comment]
Yes, that's a good idea, thanks, I didn't profile before/after the changes.

**Comment by GiacomoGarbin** [comment]
Actually the math functions could be inside the `ultra_rom` library instead of `nustd`.

In the `master` branch, these are the undefined references I get if don't link `ultra_rom`:


<details>
  <summary><b>click to expand</b></summary>

  ```
  __osPiTable
  __osMotorAccess
  alAudioFrame
  alClose
  alHeapDBAlloc
  alHeapInit
  alInit
  alLink
  alSndpAllocate
  alSndpDeallocate
  alSndpGetState
  alSndpNew
  alSndpPlay
  alSndpSetPan
  alSndpSetPitch
  alSndpSetSound
  alSndpSetVol
  alSndpStop
  alUnlink
  cosf
  guMtxCatF
  guMtxF2L
  guMtxIdent
  guMtxIdentF
  guMtxL2F
  guOrtho
  guOrthoF
  guPerspective
  guPerspectiveF
  guTranslate
  memcpy
  osAiGetLength
  osAiSetFrequency
  osAiSetNextBuffer
  osCartRomInit
  osContGetQuery
  osContGetReadData
  osContInit
  osContStartQuery
  osContStartReadData
  osCreateMesgQueue
  osCreatePiManager
  osCreateScheduler
  osCreateThread
  osEPiStartDma
  osGetIntMask
  osGetTime
  osInitialize
  osInvalDCache
  osMemSize
  osMotorInit
  osRecvMesg
  osScAddClient
  osScGetCmdQ
  osSendMesg
  osSetEventMesg
  osSetIntMask
  osSetThreadPri
  osSetTimer
  osSpTaskLoad
  osSpTaskStartGo
  osStartThread
  osTvType
  osViBlack
  osViGetCurrentFramebuffer
  osViModeTable
  osVirtualToPhysical
  osViSetEvent
  osViSetMode
  osViSetSpecialFeatures
  osWritebackDCache
  osWritebackDCacheAll
  sinf
  sprintf
  sqrtf
  strlen
  ```
</details>

And these are the ones I get if don't link `nustd` instead:

```
memset
strcpy
```

**Comment by mwpenny** [comment]
The math functions are in fact inside libultra. Sorry if that wasn't clear in the library usage table.

If you link libm first, they should be picked up from there instead of libultra.

**Comment by GiacomoGarbin** [comment]
> Sorry if that wasn't clear in the library usage table.

Oh sorry, it was clearly stated, but my brain seems to have glossed over that info.

> If you link libm first, they should be picked up from there instead of libultra.

Ah thank you! I guess this may also explain why if I link `libc` first I get a multiple definition of `__assert` coming from `/opt/crashsdk/mips64-elf/lib/libc.a` (first defined in `build/src/util/assert.o`)

**Comment by mwpenny** [comment]
The linker looks for symbols by evaluating libraries from left to right. It ignores object files in libraries which do not provide unresolved symbols (symbols are not present or were already provided by previously evaluated libraries). 

So if `assert.o` is linked before libc, the duplicate `__assert` in libc is ignored. But this isn't the case for object files _we_ define. If libc is linked first, `assert.o`'s `__assert` won't be ignored and so you get the duplicate symbol error.

**Comment by GiacomoGarbin** [comment]
Thanks a lot for the explanation, but I still don't get why this happens when what I change is just the order between `libultra` and `libc`.

This is the `ld` command that throws the multiple definition error:
<details>
<summary><b>click to expand</b></summary>

```
mips-n64-ld --no-check-sections -o build/codesegment_no_debug.o -r  build/src/audio/audio.o  build/src/audio/audiomgr.o  build/src/audio/clips.o  build/src/audio/soundarray.o  build/src/audio/soundplayer.o  build/src/controls/controller_actions.o  build/src/controls/controller.o  build/src/controls/rumble_pak.o  build/src/decor/decor_object.o  build/src/decor/decor_object_list.o  build/src/effects/effect_definitions.o  build/src/effects/effects.o  build/src/effects/portal_trail.o  build/src/effects/splash_particle_effect.o  build/src/font/dejavusans.o  build/src/font/dejavusans_images.o  build/src/font/font.o  build/src/font/liberation_mono.o  build/src/font/liberation_mono_images.o  build/src/graphics/color.o  build/src/graphics/debug_render.o  build/src/graphics/graphics.o  build/src/graphics/image.o  build/src/graphics/initgfx.o  build/src/graphics/profile_task.o  build/src/graphics/render_scene.o  build/src/graphics/renderstate.o  build/src/graphics/screen_clipper.o  build/src/levels/credits.o  build/src/levels/cutscene_runner.o  build/src/levels/intro.o  build/src/levels/level_definition.o  build/src/levels/levels.o  build/src/levels/material_state.o  build/src/levels/static_render.o  build/src/locales/locales.o  build/src/main.o  build/src/materials/light.o  build/src/materials/point_light_rendered.o  build/src/materials/shadow_caster.o  build/src/materials/subject.o  build/src/math/basis.o  build/src/math/box2d.o  build/src/math/box3d.o  build/src/math/boxs16.o  build/src/math/mathf.o  build/src/math/matrix.o  build/src/math/plane.o  build/src/math/quaternion.o  build/src/math/ray.o  build/src/math/rotated_box.o  build/src/math/transform.o  build/src/math/vector2.o  build/src/math/vector2s16.o  build/src/math/vector3.o  build/src/math/vector4.o  build/src/menu/audio_options.o  build/src/menu/cheat_codes.o  build/src/menu/confirmation_dialog.o  build/src/menu/controls.o  build/src/menu/game_menu.o  build/src/menu/gameplay_options.o  build/src/menu/joystick_options.o  build/src/menu/landing_menu.o  build/src/menu/load_game.o  build/src/menu/main_menu.o  build/src/menu/menu_builder.o  build/src/menu/menu.o  build/src/menu/new_game_menu.o  build/src/menu/options_menu.o  build/src/menu/savefile_list.o  build/src/menu/save_game_menu.o  build/src/menu/tabs.o  build/src/menu/text_manipulation.o  build/src/menu/translations.o  build/src/menu/video_options.o  build/src/physics/collision_box.o  build/src/physics/collision.o  build/src/physics/collision_capsule.o  build/src/physics/collision_cylinder.o  build/src/physics/collision_object.o  build/src/physics/collision_point.o  build/src/physics/collision_quad.o  build/src/physics/collision_scene.o  build/src/physics/collision_sphere.o  build/src/physics/contact_insertion.o  build/src/physics/contact_solver.o  build/src/physics/debug_renderer.o  build/src/physics/epa.o  build/src/physics/gjk.o  build/src/physics/line.o  build/src/physics/mesh_collider.o  build/src/physics/point_constraint.o  build/src/physics/raycasting.o  build/src/physics/rigid_body.o  build/src/physics/world.o  build/src/player/player.o  build/src/player/player_rumble_clips.o  build/src/savefile/checkpoint.o  build/src/savefile/savefile.o  build/src/savefile/scene_serialize.o  build/src/savefile/serializer.o  build/src/scene/ball.o  build/src/scene/ball_catcher.o  build/src/scene/ball_launcher.o  build/src/scene/box_dropper.o  build/src/scene/button.o  build/src/scene/camera.o  build/src/scene/clock.o  build/src/scene/door.o  build/src/scene/dynamic_render_list.o  build/src/scene/dynamic_scene.o  build/src/scene/elevator.o  build/src/scene/fizzler.o  build/src/scene/hud.o  build/src/scene/pedestal.o  build/src/scene/point_light.o  build/src/scene/portal.o  build/src/scene/portal_gun.o  build/src/scene/portal_render.o  build/src/scene/portal_surface.o  build/src/scene/portal_surface_generator.o  build/src/scene/portal_surface_gfx.o  build/src/scene/render_plan.o  build/src/scene/scene_animator.o  build/src/scene/scene.o  build/src/scene/security_camera.o  build/src/scene/shadow_map.o  build/src/scene/shadow_renderer.o  build/src/scene/signage.o  build/src/scene/signals.o  build/src/scene/static_scene.o  build/src/scene/switch.o  build/src/scene/trigger_listener.o  build/src/sk64/skelatool_animator.o  build/src/sk64/skelatool_armature.o  build/src/util/assert.o  build/src/util/dynamic_animated_asset_data.o  build/src/util/dynamic_asset_data.o  build/src/util/dynamic_asset_loader.o  build/src/util/linked_list.o  build/src/util/memory.o  build/src/util/profile.o  build/src/util/rom.o  build/src/util/time.o build/assets/models/player/chell_geo.o build/assets/models/portal_gun/ball_trail_geo.o build/assets/models/portal_gun/v_portalgun_geo.o build/assets/models/portal_gun/w_portalgun_geo.o build/assets/models/props/round_elevator_geo.o build/assets/models/props/round_elevator_interior_geo.o build/assets/models/props/round_elevator_collision_geo.o build/assets/models/props/signage_geo.o build/assets/models/portal/portal_blue_geo.o build/assets/models/portal/portal_blue_filled_geo.o build/assets/models/portal/portal_blue_face_geo.o build/assets/models/portal/portal_collider_geo.o build/assets/models/portal/portal_collider_vertical_geo.o build/assets/models/portal/portal_orange_geo.o build/assets/models/portal/portal_orange_filled_geo.o build/assets/models/portal/portal_orange_face_geo.o build/assets/models/grav_flare_geo.o build/assets/models/fleck_ash2_geo.o build/assets/materials/static_mat.o build/assets/materials/ui_mat.o build/assets/materials/hud_mat.o build/src/audio/subtitles.o build/src/audio/languages.o -L/opt/crashsdk/mips64-elf/lib -l:libc.a -l:libm.a -L/usr/lib/n64 -lultra_rom -L/opt/crashsdk/lib/gcc/mips64-elf/12.2.0 -lgcc

mips-n64-ld: /opt/crashsdk/mips64-elf/lib/libc.a(libc_a-assert.o): in function `__assert':
(.text+0x5c): multiple definition of `__assert'; build/src/util/assert.o:(.text+0x0): first defined here
```
</details>

While this is the "normal" `ld` command:
<details>
<summary><b>click to expand</b></summary>

```
mips-n64-ld --no-check-sections -o build/codesegment_no_debug.o -r  build/src/audio/audio.o  build/src/audio/audiomgr.o  build/src/audio/clips.o  build/src/audio/soundarray.o  build/src/audio/soundplayer.o  build/src/controls/controller_actions.o  build/src/controls/controller.o  build/src/controls/rumble_pak.o  build/src/decor/decor_object.o  build/src/decor/decor_object_list.o  build/src/effects/effect_definitions.o  build/src/effects/effects.o  build/src/effects/portal_trail.o  build/src/effects/splash_particle_effect.o  build/src/font/dejavusans.o  build/src/font/dejavusans_images.o  build/src/font/font.o  build/src/font/liberation_mono.o  build/src/font/liberation_mono_images.o  build/src/graphics/color.o  build/src/graphics/debug_render.o  build/src/graphics/graphics.o  build/src/graphics/image.o  build/src/graphics/initgfx.o  build/src/graphics/profile_task.o  build/src/graphics/render_scene.o  build/src/graphics/renderstate.o  build/src/graphics/screen_clipper.o  build/src/levels/credits.o  build/src/levels/cutscene_runner.o  build/src/levels/intro.o  build/src/levels/level_definition.o  build/src/levels/levels.o  build/src/levels/material_state.o  build/src/levels/static_render.o  build/src/locales/locales.o  build/src/main.o  build/src/materials/light.o  build/src/materials/point_light_rendered.o  build/src/materials/shadow_caster.o  build/src/materials/subject.o  build/src/math/basis.o  build/src/math/box2d.o  build/src/math/box3d.o  build/src/math/boxs16.o  build/src/math/mathf.o  build/src/math/matrix.o  build/src/math/plane.o  build/src/math/quaternion.o  build/src/math/ray.o  build/src/math/rotated_box.o  build/src/math/transform.o  build/src/math/vector2.o  build/src/math/vector2s16.o  build/src/math/vector3.o  build/src/math/vector4.o  build/src/menu/audio_options.o  build/src/menu/cheat_codes.o  build/src/menu/confirmation_dialog.o  build/src/menu/controls.o  build/src/menu/game_menu.o  build/src/menu/gameplay_options.o  build/src/menu/joystick_options.o  build/src/menu/landing_menu.o  build/src/menu/load_game.o  build/src/menu/main_menu.o  build/src/menu/menu_builder.o  build/src/menu/menu.o  build/src/menu/new_game_menu.o  build/src/menu/options_menu.o  build/src/menu/savefile_list.o  build/src/menu/save_game_menu.o  build/src/menu/tabs.o  build/src/menu/text_manipulation.o  build/src/menu/translations.o  build/src/menu/video_options.o  build/src/physics/collision_box.o  build/src/physics/collision.o  build/src/physics/collision_capsule.o  build/src/physics/collision_cylinder.o  build/src/physics/collision_object.o  build/src/physics/collision_point.o  build/src/physics/collision_quad.o  build/src/physics/collision_scene.o  build/src/physics/collision_sphere.o  build/src/physics/contact_insertion.o  build/src/physics/contact_solver.o  build/src/physics/debug_renderer.o  build/src/physics/epa.o  build/src/physics/gjk.o  build/src/physics/line.o  build/src/physics/mesh_collider.o  build/src/physics/point_constraint.o  build/src/physics/raycasting.o  build/src/physics/rigid_body.o  build/src/physics/world.o  build/src/player/player.o  build/src/player/player_rumble_clips.o  build/src/savefile/checkpoint.o  build/src/savefile/savefile.o  build/src/savefile/scene_serialize.o  build/src/savefile/serializer.o  build/src/scene/ball.o  build/src/scene/ball_catcher.o  build/src/scene/ball_launcher.o  build/src/scene/box_dropper.o  build/src/scene/button.o  build/src/scene/camera.o  build/src/scene/clock.o  build/src/scene/door.o  build/src/scene/dynamic_render_list.o  build/src/scene/dynamic_scene.o  build/src/scene/elevator.o  build/src/scene/fizzler.o  build/src/scene/hud.o  build/src/scene/pedestal.o  build/src/scene/point_light.o  build/src/scene/portal.o  build/src/scene/portal_gun.o  build/src/scene/portal_render.o  build/src/scene/portal_surface.o  build/src/scene/portal_surface_generator.o  build/src/scene/portal_surface_gfx.o  build/src/scene/render_plan.o  build/src/scene/scene_animator.o  build/src/scene/scene.o  build/src/scene/security_camera.o  build/src/scene/shadow_map.o  build/src/scene/shadow_renderer.o  build/src/scene/signage.o  build/src/scene/signals.o  build/src/scene/static_scene.o  build/src/scene/switch.o  build/src/scene/trigger_listener.o  build/src/sk64/skelatool_animator.o  build/src/sk64/skelatool_armature.o  build/src/util/assert.o  build/src/util/dynamic_animated_asset_data.o  build/src/util/dynamic_asset_data.o  build/src/util/dynamic_asset_loader.o  build/src/util/linked_list.o  build/src/util/memory.o  build/src/util/profile.o  build/src/util/rom.o  build/src/util/time.o build/assets/models/player/chell_geo.o build/assets/models/portal_gun/ball_trail_geo.o build/assets/models/portal_gun/v_portalgun_geo.o build/assets/models/portal_gun/w_portalgun_geo.o build/assets/models/props/round_elevator_geo.o build/assets/models/props/round_elevator_interior_geo.o build/assets/models/props/round_elevator_collision_geo.o build/assets/models/props/signage_geo.o build/assets/models/portal/portal_blue_geo.o build/assets/models/portal/portal_blue_filled_geo.o build/assets/models/portal/portal_blue_face_geo.o build/assets/models/portal/portal_collider_geo.o build/assets/models/portal/portal_collider_vertical_geo.o build/assets/models/portal/portal_orange_geo.o build/assets/models/portal/portal_orange_filled_geo.o build/assets/models/portal/portal_orange_face_geo.o build/assets/models/grav_flare_geo.o build/assets/models/fleck_ash2_geo.o build/assets/materials/static_mat.o build/assets/materials/ui_mat.o build/assets/materials/hud_mat.o build/src/audio/subtitles.o build/src/audio/languages.o -L/usr/lib/n64 -lultra_rom -L/opt/crashsdk/mips64-elf/lib -l:libc.a -l:libm.a -L/opt/crashsdk/lib/gcc/mips64-elf/12.2.0 -lgcc
```
</details>

The only difference is the order between `libultra` and `libc` at the very end, and in both commands `assert.o` appears before both libraries.

**Comment by GiacomoGarbin** [comment]
I _think_ I got it now.

> The linker [...] ignores <ins>object files</ins> in libraries which do not provide unresolved symbols

When the linker first scans `libc` for unresolved symbols, it eventually also picks up `libc_a-assert.o` (perhaps to resolve some unresolved symbols generated by `libc` itself) while when `libultra` is scanned first, most unresolved symbols are resolved, and the remaining unresolved symbols (if any) do not require adding `libc_a-assert.o` during the following `libc` scan.

File: Makefile [hunk 1]
```diff
@@ -79,7 +79,7 @@ LCDEFS +=	-DF3DEX_GBI_2 -DSCENE_SCALE=${SCENE_SCALE}
 
 LDIRT  =	$(BASE_TARGET_NAME).elf $(CP_LD_SCRIPT) $(BASE_TARGET_NAME).z64 $(BASE_TARGET_NAME)_no_debug.map $(ASMOBJECTS)
 
-LDFLAGS =	-L$(N64_ROOT)/usr/lib/n64 $(N64LIB)  -L$(N64_LIBGCCDIR) -lgcc
+LDFLAGS =	-L$(N64_ROOT)/usr/lib/n64 $(N64LIB) -L$(N64_ROOT)/opt/crashsdk/mips64-elf/lib -l:libc.a -l:libm.a -L$(N64_LIBGCCDIR) -lgcc
```
Comment (mwpenny) [inline Makefile:82]: `-lc` and `-lm` should suffice.

Also can you make the location of the include and library paths configurable via a Makefile variable?

Comment (mwpenny) [inline Makefile:82]: Linking libm after libultra has no effect since libultra contains the math functions.

**Comment by mwpenny** [comment]
Yes, that makes sense.

When you specify libc first, the linker will get `strlen()`, `sprintf()`, and so on from there instead of libultra. It looks like the libc implementations depend on symbols defined in libc's `assert.o` which conflicts with Portal 64's 

I had some time today and got this built on my Linux machine but had issues on Windows (MSYS2 environment). I plan to rework the build system soon to make it more flexible. I did some makefile changes previously toward this goal and want to make sure the build continues to work in several different environments with your changes. It's probably a problem on my end, so I'll need to play around a bit more. I left some comments in the meantime.

**Comment by GiacomoGarbin** [comment]
Thanks for the code comments, [8bed9b6](https://github.com/mwpenny/portal64-still-alive/pull/47/commits/8bed9b623852e402cde8105ada11be1fd58de949) should address the issues. And sorry for the conflicts in makefile.

**Comment by mwpenny** [comment]
The Windows-only build issues were indeed due to a problem with my environment.

Also sorry for any confusion about libm. What I meant by my comment was that for it to be of use it needs to be linked before libultra. Even though libultra is still required right now, it's good to gradually remove dependencies on it like you're doing.

I looked into usage of `sinf()` and `cosf()`. There are only a few places where they are called:

1. Some level object initialization on scene creation
2. Some unused functions
3. In the following locations once per frame (worst case):
    * `playerUpdate()` calls `quatAxisAngle()` twice, which calls `sinf()` and `cosf()` once each
    * `pointConstraintRotateTo()` is called by the contact solver once per point constraint, which in turn calls `quatDecompose()` once, which calls `sinf()` once. Right now there can only be one point constraint (for the currently grabbed object)

1 and 2 don't matter - so at most 5 calls. This shouldn't have a noticeable effect, even if/when more game elements are added or the low-level functions that call `sinf()` and `cosf()` are called more frequently. I was concerned about heavy usage and am happy to be wrong.

I was unable to build with libm though on Linux or Windows. Were you? I get
> small-data section exceeds 64KB; lower small-data size limit (see option -G)
> [...]
> build/codesegment_no_debug.o: in function '__errno':
> (.text+0x433b8): relocation truncated to fit: R_MIPS_GPREL16 against `_impure_ptr'

and similar. Even though `-G 0` is being used.

If you also have this problem, I'm good to merge your removal of libnustd and solve the libm issue / replace the math functions in a future PR.

**Comment by GiacomoGarbin** [comment]
Ah sorry for misunderstanding your comment about libm and thanks for the clarification.

Thanks also for looking into sinf/cosf usage. I wanted to mention this already in my last comment, and you're probably already aware of it, since you mentioned it in your first comment, pretty much every time I come across some document/video about trigonometric functions on N64, it's advised against using those of the standard library for better performance. libdragon also provides an optimized version of sinf/cosf, although they were implemented for graphics computations, so I'm not sure if they are ideal for Portal usage.

I tried linking libm before libultra and get the same error as you (plus a few undefined references):

<details>
<summary><b>click to expand</b></summary>

```
mips-n64-ld --no-check-sections -L. -T build/portal_no_debug.ld -Map build/portal_no_debug.map -o build/portal.elf
mips-n64-ld: small-data section exceeds 64KB; lower small-data size limit (see option -G)
build/codesegment_no_debug.o: in function `__errno':
(.text+0x433b8): relocation truncated to fit: R_MIPS_GPREL16 against `_impure_ptr'
mips-n64-ld: build/codesegment_no_debug.o: in function `_close_r':
(.text+0x439d4): undefined reference to `close'
build/codesegment_no_debug.o: in function `_reclaim_reent':
(.text+0x43a08): relocation truncated to fit: R_MIPS_GPREL16 against `_impure_ptr'
mips-n64-ld: build/codesegment_no_debug.o: in function `_lseek_r':
(.text+0x43b08): undefined reference to `lseek'
mips-n64-ld: build/codesegment_no_debug.o: in function `_read_r':
(.text+0x43b60): undefined reference to `read'
mips-n64-ld: build/codesegment_no_debug.o: in function `_write_r':
(.text+0x43bb8): undefined reference to `write'
build/codesegment_no_debug.o: in function `fclose':
(.text+0x447d0): relocation truncated to fit: R_MIPS_GPREL16 against `_impure_ptr'
build/codesegment_no_debug.o: in function `fflush':
(.text+0x44a80): relocation truncated to fit: R_MIPS_GPREL16 against `_impure_ptr'
mips-n64-ld: build/codesegment_no_debug.o: in function `_sbrk_r':
(.text+0x44aa0): undefined reference to `sbrk'
```
</details>

Is it possible that what is missing the `-G0` option is the libm we are linking?

**Comment by mwpenny** [comment]
> Is it possible that what is missing the -G0 option is the libm we are linking?

I tried with the crashsdk newlib and also built it from source using `-G0` and had the same errors :(

> [...] advised against using those of the standard library for better performance.

You're absolutely right, I had the same thing in mind. What's nice here is that the game's usage of the functions low and so in theory the impact should be as well.

But since only a handful of standard library functions are needed anyway and they're relatively simple (except `sprintf()` - but it's only used to convert integers to strings), I'm actually leaning toward eventually reimplementing/replacing those few functions and avoiding the newlib dependency entirely until proven necessary. That would give more control over performance and also simplify the build, avoiding issues like we each had (thanks for confirming on your end). What you have here is a good change regardless of what happens later. One less proprietary dependency :)

Libdragon's implementation of the trig functions is likely faster than newlib's for N64 and it's definitely worth considering since I want to use libdragon anyway. There's bit more work needed before starting to integrate libdragon though (I will start #4 soon as a stepping stone). With all of this in mind and the fact we have these build issues, I think the best path forward is to merge your changes to remove the libnustd dependency and handle the trig functions later - whether that results in reimplementation or using libdragon's.

**Comment by mwpenny** [comment]
Thanks!

It seems like your interest right now with this project is refactoring to remove proprietary code requirements. If you're looking for something more to contribute in this area, it would be helpful to start on a library-independent interface for one of the simpler APIs - for example, the controller.

This could serve as a starting point to test libultra integration. For each point of abstraction, I was thinking of having one header and multiple implementation C files (libultra-based, libdragon-based, etc.). The C file to link against would be chosen at build time. We don't need to do this part yet though - the first step is working on the abstraction to validate and decide on an approach.

It's more of an investigation than a feature. If you decide to work on this, just create an issue first. No worries if you decide not to.

**Comment by GiacomoGarbin** [comment]
> I tried with the crashsdk newlib and also built it from source using `-G0`

Can I ask out of curiosity (and to learn something new) how did you recompile libm?

> I'm actually leaning toward eventually reimplementing/replacing those few functions and avoiding the newlib dependency entirely until proven necessary.

It is definitely a valid option to consider.

> Thanks!

Thank you for managing/leading the project and for the opportunity to collaborate!

> It seems like your interest right now with this project is refactoring to remove proprietary code requirements.

You're right, I'm interested in working in this area of the project. :)

> For each point of abstraction, I was thinking of having one header and multiple implementation C files (libultra-based, libdragon-based, etc.). The C file to link against would be chosen at build time.

This is exactly how I was approaching the issue too.

I started by looking into the controller as you suggested and abstracted `controls/controller.h` from libultra, luckily `controls/controller_action.h` and `controls/rumble_pack.h` were already libultra independent. Now I can proceed to abstract more parts of the game or attempt to add a libdragon implementation for the controller .c files. The second option looks tricky because the controller logic runs on its own thread and I don't know yet how it would be handled in libdragon, especially while keeping the rest of the game tied to libultra. But sooner or later we will have to face this problem.

> If you decide to work on this, just create an issue first.

Okay I'll open the issue, should it only be about the controller part or the whole libdragon transition? If I want to share what I have so far, should I just push to my forked repo and link the commits, instead of opening the pull request?

**Comment by mwpenny** [comment]
> Can I ask out of curiosity (and to learn something new) how did you recompile libm?

You can download the source code from https://sourceware.org/newlib/. Then it's just a matter of cross-compiling by calling the `configure` script with the proper arguments and then running `make`. Here's an example: https://n64brew.dev/wiki/Building_GCC#Building_Newlib.

> I started by looking into the controller as you suggested and abstracted controls/controller.h from libultra
> [...]
> I don't know yet how it would be handled in libdragon

Great to hear! We're not quite ready for libdragon just yet. I will update the build system to support it (and switching between) in #4, after chamber 15 is finished. You're right that there are core elements like scheduling which touch everything.

For the time being, the important part is refactoring to have points of abstraction even if there is only one implementation. How about an issue for the controller code + a draft PR to review, and we go from there? Making it a PR makes review and discussion easier, but marking as a draft signifies it's still a work in progress.

**Comment by GiacomoGarbin** [comment]
> You can download the source code from https://sourceware.org/newlib/. Then it's just a matter of cross-compiling by calling the configure script with the proper arguments and then running make. Here's an example: https://n64brew.dev/wiki/Building_GCC#Building_Newlib.

Thanks for the explanation and the link!

> For the time being, the important part is refactoring to have points of abstraction even if there is only one implementation. How about an issue for the controller code + a draft PR to review, and we go from there?

Ok thanks, I'll open the issue for the controller and the draft PR, and then move to the next area to abstract.

**Comment by mwpenny** [comment]
Nice work! I took a look and left some comments.

As for what's next, did you have a particular area in mind? I generally prefer opening an issue first to be helpful since it can provide an opportunity to share ideas and/or clarify things at the start and make the implementation and review easier.

Some ideas that come to mind:
* Time-related functions
  * `osGetTime()` and `OSTimer`
  * Note that the game's `OSTimer` usage is only for creating a delay - it is effectively just an asynchronous `sleep()` and can be wrapped as such
* Standard integer types (you already did this for the controller)
* Video/screen initialization

No worries about the Docker image! Eventually CI builds on push (via GitHub actions) would be nice to add but I haven't looked at how difficult this would be to implement while not exposing Portal game files or the compiled ROM. Might need to self-host it. Not sure if you're also interested in looking at available options here.

**Comment by GiacomoGarbin** [comment]
> Nice work! I took a look and left some comments.

Thanks!

> As for what's next, did you have a particular area in mind?

I had started looking at files in the `src/math` folder where only `matrix.h` and `transform.h` need refactoring. I can finish what I started and then move on to your suggestions, if that's okay?

> For open-ended items I generally prefer opening an issue first since it can provide an opportunity to share ideas and/or clarify things at the start and make the implementation and review easier.

Okay, I'll open the issue first.

> Eventually CI builds on push (via GitHub actions) would be nice to add but I haven't looked at how difficult this would be to implement while not exposing Portal game files or the compiled ROM. Might need to self-host it. Not sure if you're also interested in looking at available options here.

Having CI would be nice and very useful, maybe we can open an issue for that too? I took a quick look at the runners hosted on GitHub and yes, the main issue, as you pointed out, seems to be keeping game data private. Maybe (I'm just making a guess) the Github CI artefacts aren't exposed publicly and I think (but I'm not 100% sure) that Steam has an API that allows you to download data, but you should still provide Steam credentials without hardcoding them in the Actions script since the script is part of the public repository. Also, downloading data with each workflow will increase build time, and logging into Steam may still require manual authentication (e.g. two-factor authentication). A self-hosted runner should overcome all these limitations. I personally don't have the capability of hosting a server, maybe there are some 3rd party free solutions available?

**Comment by mwpenny** [comment]
> I had started looking at files in the src/math folder
I left a comment on #62 to continue the discussion this there. Thanks for opening the issue.

> Maybe we can open an issue for that too

Good idea! I've opened #63 to continue the discussion there. I've captured our ideas from this conversation and described what I've found after some cursory research. It looks like it may be possible with SteamCMD and GitHub-hosted runners, but should be confirmed.

There is [SteamCMD](https://developer.valvesoftware.com/wiki/SteamCMD) which can be used to download game files. Credentials can be stored using GitHub Actions [secrets](https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions), but the open question here is whether or not these files would be publicly accessible after a workflow run.

**Pull request merged**